nixos/pam: Move `pam_fprintd.so` after `pam_unix.so`

skeleten opened the pull request:
This enables logging into GUI applications using a password even when
using `fprintd`. However, this now requires first entering an empty
password before being prompted for a fingerprint.

Fixes #171136

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This moves the checking of the `pam_fprintd` module *after* the `pam_unix` module. This is required for GUI applications to still be authenticated with a password

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius commented on nixos/modules/security/pam.nix:
```diff
@@ -542,6 +539,9 @@ let
           optionalString cfg.unixAuth ''
             auth sufficient pam_unix.so ${optionalString cfg.allowNullPassword "nullok"} ${optionalString cfg.nodelay "nodelay"} likeauth try_first_pass
+          '' +
+                    optionalString cfg.fprintAuth ''
```
formatting looks weird

RaitoBezarius replied:
this was not addressed.

RaitoBezarius reviewed (changes requested):
Address current comments please.

rht commented:
Bump. The formatting issue has been addressed. Still needs a rebase though.

> However, this now requires first entering an empty
password before being prompted for a fingerprint.

I think the current behavior also requires you to press enter, which is equivalent to entering an empty password anyway.

jtojnar commented:
> I think the current behavior also requires you to press enter

At least in GDM and GNOME Shell polkit agent, you just put your finger on the fingerprint reader after being prompted, and it authenticates you without any further action.

rht commented:
Ah, right, I forgot to mention that I'm using i3 & i3lock.

RaitoBezarius commented:
I am on vacations so I cannot touch this yet, I recommend to rebase and look for other reviewers on Discourse.

skeleten committed:
nixos/pam: Move `pam_fprintd.so` after `pam_unix.so`

This enables logging into GUI applications using a password even when
using `fprintd`. However, this now requires first entering an empty
password before being prompted for a fingerprint.

Fixes #171136

nyabinary commented:
Needs a rebase

jtojnar reviewed (changes requested):
> However, this now requires first entering an empty
password before being prompted for a fingerprint.

This is a downgrade IMO.

Maybe something like https://github.com/NixOS/nixpkgs/issues/171136#issuecomment-1627779037 would work better.

nyabinary commented:
> > However, this now requires first entering an empty
> > password before being prompted for a fingerprint.
> 
> This is a downgrade IMO.
> 
> Maybe something like [#171136 (comment)](https://github.com/NixOS/nixpkgs/issues/171136#issuecomment-1627779037) would work better.

I mean, no one made a config PR yet sadly...

jtojnar commented:
> I mean, no one made a config PR yet sadly...

That does not mean we should merge a broken PR instead.

And someone actually made a PR with what looked like a proper fix but there was still a bug and we lack people familiar with PAM to review it: https://github.com/NixOS/nixpkgs/pull/282322

AkechiShiro commented:
> > I think the current behavior also requires you to press enter
> 
> At least in GDM and GNOME Shell polkit agent, you just put your finger on the fingerprint reader after being prompted, and it authenticates you without any further action.

This may be related to https://github.com/sddm/sddm/pull/1220#issuecomment-1829545241

kcoderhtml commented:
would making an option in the fprint service to have password auth first be a better solution? it would be backwards compatible and still allow people to use the password first method without hacky workarounds
